addons/l10n_sa_edi/models/account_journal.py

```diff
@@ -0,0 +1,256 @@
+        edi_format = self.env.ref('l10n_sa_edi.edi_sa_zatca')
+        CCSID_data = json.loads(self.l10n_sa_compliance_csid_json)
+        compliance_files = self._l10n_sa_get_compliance_files()
+        for fname, fval in compliance_files.items():
+            invoice_hash_hex = edi_format._l10n_sa_generate_invoice_xml_hash(fval).decode()
+            digital_signature = edi_format._l10n_sa_get_digital_signature(self.company_id, invoice_hash_hex).decode()
+            prepared_xml = self._l10n_sa_prepare_compliance_xml(fname, fval, CCSID_data['binarySecurityToken'],
+                                                                digital_signature)
+            result = self._l10n_sa_api_compliance_checks(prepared_xml.decode(), CCSID_data)
+            if result.get('error'):
+                raise UserError(_("Could not complete Compliance Checks for the following file: %s") % fname)
+            if result['validationResults']['status'] != 'PASS':
+                raise UserError(_("Could not complete Compliance Checks for the following file: %s") % fname)
+        self.l10n_sa_compliance_checks_passed = True
+
+    def _l10n_sa_prepare_compliance_xml(self, xml_name, xml_raw, PCSID, signature):
+        """
+            Prepare XML content to be used for Compliance checks
+        :param xml_name: Name of the compliance file to be used
+        :param xml_raw: Raw content of the compliance file
+        :param PCSID: X509 certificate obtained for the journal
+        :return: prepared & signed xml content for compliance checks
+        """
+        xml_content = self._l10n_sa_prepare_invoice_xml(xml_raw)
+        signed_xml = self._l10n_sa_sign_xml(xml_content, PCSID, signature)
+        return signed_xml
+
+    def _l10n_sa_prepare_invoice_xml(self, xml_content):
+        """
+            Prepare the XML content of the test invoices before running the compliance checks
+        :param xml_content:
+        :return: processed string of the provided XML content
+        """
+        ubl_extensions = etree.fromstring(self.env.ref('l10n_sa_edi.export_sa_zatca_ubl_extensions')._render())
+        root = etree.fromstring(xml_content.encode())
+        root.insert(0, ubl_extensions)
+        ns_map = self._l10n_sa_get_edi_format()._l10n_sa_get_namespaces()
+
+        def _get_node(xpath_str):
+            return root.xpath(xpath_str, namespaces=ns_map)[0]
+
+        # Update the Company VAT number in the test invoice
+        vat_el = _get_node('//cbc:CompanyID')
+        vat_el.text = self.company_id.vat
+
+        # Update the Company Name in the test invoice
+        name_nodes = ['cac:PartyName/cbc:Name', 'cac:PartyLegalEntity/cbc:RegistrationName', 'cac:Contact/cbc:Name']
+        for node in name_nodes:
+            comp_name_el = _get_node('//cac:AccountingSupplierParty/cac:Party/' + node)
+            comp_name_el.text = self.company_id.display_name
+
+        return etree.tostring(root)
+
+    def _l10n_sa_can_submit_einvoices(self):
```

Comment:
```suggestion
    def _l10n_sa_ready_to_submit_einvoices(self):
``` The method name could have another meaning  (I would think check if the edi format is set or something)

Reply:
Updated